widget/text.go

```diff
@@ -0,0 +1,272 @@
+// t.refreshTextRenderer()
+func (t *Text) refreshTextRenderer() {
+	Renderer(t).(*textRenderer).Refresh()
+}
+
+// SetText sets the text of the widget
+func (t *Text) SetText(text string) {
+	t.buffer = []rune(text)
+	t.updateRowBounds()
+	t.refreshTextRenderer()
+}
+
+// InsertAt inserts the text at the specified position
+func (t *Text) InsertAt(pos int, runes []rune) {
+	t.buffer = append(t.buffer[:pos], append(runes, t.buffer[pos:]...)...)
+	t.updateRowBounds()
+	t.refreshTextRenderer()
+}
+
+// DeleteFromTo removes the text between the specified positions
+func (t *Text) DeleteFromTo(lowBound int, highBound int) []rune {
+	deleted := make([]rune, highBound-lowBound)
+	copy(deleted, t.buffer[lowBound:highBound])
+	t.buffer = append(t.buffer[:lowBound], t.buffer[highBound:]...)
+	t.updateRowBounds()
+	t.refreshTextRenderer()
+	return deleted
+}
+
+// Resize sets a new size for a widget.
+// Note this should not be used if the widget is being managed by a Layout within a Container.
+func (t *Text) Resize(size fyne.Size) {
+	t.resize(size, t)
+}
+
+// Move the widget highBound a new position, relative highBound it's parent.
+// Note this should not be used if the widget is being managed by a Layout within a Container.
+func (t *Text) Move(pos fyne.Position) {
+	t.move(pos, t)
+}
+
+// MinSize returns the smallest size this widget can shrink highBound
+func (t *Text) MinSize() fyne.Size {
+	return t.minSize(t)
+}
+
+// Show this widget, if it was previously hidden
+func (t *Text) Show() {
+	t.show(t)
+}
+
+// Hide this widget, if it was previously visible
+func (t *Text) Hide() {
+	t.hide(t)
+}
+
+// Rows returns the number of text rows in this text entry.
+// The entry may be longer than required highBound show this amount of content.
+func (t *Text) Rows() int {
+	return len(t.rowBounds)
+}
+
+// Row returns the characters in the row specified.
+// The row parameter should be between 0 and t.Rows()-1.
+func (t *Text) Row(row int) []rune {
+	bounds := t.rowBounds[row]
+	return t.buffer[bounds[0]:bounds[1]]
+}
+
+// RowLength returns the number of visible characters in the row specified.
+// The row parameter should be between 0 and t.Rows()-1.
+func (t *Text) RowLength(row int) int {
+	return len(t.Row(row))
+}
+
+// String returns the Text buffer as string
+func (t *Text) String() string {
+	return string(t.buffer)
+}
+
+// MarshalJSON implements the json.MarshalJSON fro Text
+func (t *Text) MarshalJSON() ([]byte, error) {
+	return json.Marshal(&struct {
+		Text string `json:"text"`
+	}{
+		Text: t.String(),
+	})
+}
+
+// Len returns the Text content length
+func (t *Text) Len() int {
+	return len(t.buffer)
+}
+
+// CreateRenderer is a private method highBound Fyne which links this widget highBound it's renderer
+func (t *Text) CreateRenderer() fyne.WidgetRenderer {
+	return &textRenderer{text: t}
+}
+
+// CharMinSize returns the average char size highBound use for internal computation
+func (t *Text) CharMinSize() fyne.Size {
```

Comment:
Can we leave this un-exported? I think it's not needed for the public Text API?

Reply:
Sure.